### PR TITLE
Kernel/Storage: Modernize ATA code

### PR DESCRIPTION
**Pull request by supercomputer7**
The main goal of this pull request is to modernize the ATA code so we can support more ATA controllers in the near future (e.g. Intel PIIX3 controllers, Intel PIIX4-compatible ATA controllers, Intel ICH0-3 IDE and Intel ICH5-9 SATA controllers).

It's also possible to support IDE controllers not from Intel as a vendor, but I personally don't own any machine that has them. I do know that VIA and ATI/AMD IDE controllers were common in the past, so someone can take this framework (the inspiration for the API is from libata of the Linux kernel) and apply this on other controllers as long as these controllers conform to the ATA spec.

There's still plenty of things to do, this is just the start :)

### CODE REVIEW

File: Kernel/Storage/ATA/ATAPort.cpp
Comment (Panky-codes) [inline Kernel/Storage/ATA/ATAPort.cpp:169]: I think this should not be a comment anymore? I did have a deadlock situation in NVMe driver without unlocking the lock before completing the request. 
Same applies to complete request at the end

File: Kernel/Storage/ATA/GenericIDE/Channel.cpp
Comment (Panky-codes) [inline Kernel/Storage/ATA/GenericIDE/Channel.cpp:103]: As we have ErrorOr return type, we should change these to `TRY` return the error

File: Kernel/Storage/ATA/ATAPort.cpp
Comment (Panky-codes) [inline Kernel/Storage/ATA/ATAPort.cpp:163]: Can't we just complete the AsyncDeviceRequest with an error here if the `result` is an error?

File: Kernel/Storage/ATA/GenericIDE/Channel.cpp
Comment (supercomputer7) [reply Kernel/Storage/ATA/GenericIDE/Channel.cpp:103]: It's merely about merging both files together, so I prefer to do that in another commit (or maybe even in another PR).

File: Kernel/Storage/ATA/ATAPort.cpp
Comment (supercomputer7) [reply Kernel/Storage/ATA/ATAPort.cpp:163]: Yes we can actually do that and I'll propagate it to the entire code where possible :)

**Comment by supercomputer7** [comment]
I wrote basic support for QEMU for ICH6/ICH7 IDE controllers so we can use that to implement support in our kernel too.
To enable this feature in QEMU one needs to fetch patches from here:
https://github.com/supercomputer7/qemu/tree/ich6-ide
and compile SeaBIOS to support the new controller here:
https://github.com/supercomputer7/seabios/tree/ata-piix-ich6-setup

I expect this pull request to allow us to write support for ICH6/7 PCI-native IDE controllers soon :)

**Comment by supercomputer7** [comment]
The recent tests failed because of the spinlock we hold while enumerating the PCI devices list.
https://github.com/SerenityOS/serenity/runs/5069683525?check_suite_focus=true
To fix this we need to first merge #12282.

**Comment by supercomputer7** [comment]
I'll test these patches on my ICH7 test machine and I might consider to test this on a laptop I acquired recently (ICH8 chipset, the vendor is Dell).

**Comment by supercomputer7** [comment]
The good news is that it boots OK on my ICH7 test machine. The bad news is that we still can't boot with IDE PCI native controllers. I suppose the ICH6-IDE patch I created for QEMU could help with this problem so we can debug this easily :)

**Comment by supercomputer7** [comment]
> The good news is that it boots OK on my ICH7 test machine. The bad news is that we still can't boot with IDE PCI native controllers. I suppose the ICH6-IDE patch I created for QEMU could help with this problem so we can debug this easily :)

The good news about the bad news is that I'm able to reproduce the bug I have on my ICH7 test machine on QEMU with the ICH6-IDE patch together with the SeaBIOS patch to support that device in QEMU, so it's probably a matter of fixing it in QEMU and will work in real hardware too :)

**Comment by supercomputer7** [comment]
> > The good news is that it boots OK on my ICH7 test machine. The bad news is that we still can't boot with IDE PCI native controllers. I suppose the ICH6-IDE patch I created for QEMU could help with this problem so we can debug this easily :)
> 
> The good news about the bad news is that I'm able to reproduce the bug I have on my ICH7 test machine on QEMU with the ICH6-IDE patch together with the SeaBIOS patch to support that device in QEMU, so it's probably a matter of fixing it in QEMU and will work in real hardware too :)

Apparently, when the PCI IDE controller is operated in native mode, the specification says that we should read the status (alternate register) with an offset of 2. Other than that, we should adapt the code to work with PCI IRQ routing on IOAPIC, but this is a thing we can implement in a different PR :)
Therefore, I can consider this mergeable :)

**Comment by supercomputer7** [comment]
Besides the fact that I need to rebase this, I don't like how the AHCI code looks now and I will probably adapt it to the standard of this PR now. So for now, this is a draft, again.

**Comment by supercomputer7** [comment]
This now relies on #13416 to be merged first. I intend to adapt the AHCI code to work with the ATAPort abstraction too now :)

**Comment by supercomputer7** [comment]
This is not stale, as it relies on #13416.

**Comment by supercomputer7** [comment]
This is still waiting for #13416.

**Comment by supercomputer7** [comment]
Rebased to resolve conflicts with master, still waiting for #13416 to be merged. After this is merged, we can go re-implementing AHCI code to match what we do here and to add nice features like reading SMART data and so on.

**Comment by supercomputer7** [comment]
Not stale, waiting for #13416 to be merged.

File: Kernel/Storage/ATA/ATAPort.cpp
Comment (linusg) [inline Kernel/Storage/ATA/ATAPort.cpp:89]: 😄

**Review by linusg — APPROVED** [review]
Not an expert in any of this, but LGTM so let's finally get this merged. Thanks for keeping the PR alive for so long ❤️

**Pull request merged**
